File: src/javaprograms/LearnBinarySearchAlgo.java
package javaprograms;

public class LearnBinarySearchAlgo extends GuessGame {

    private static boolean isBadVersion(int m) {
        // This is a placeholder for the actual implementation of the isBadVersion API.
        // In a real scenario, this would be provided by the system or defined elsewhere in the codebase.
        return m >= 4; // Example: versions 4 and above are bad
    }

    public LearnBinarySearchAlgo(int n) {
        super(n);
    }

    public static void main(String[] args) {

        System.out.println("Learn Binary Search Algorithm in Java");

        // ---- SQRT ----
        System.out.println("Square root of 16: " + sqrt(16));

        // ---- GUESS NUMBER ----
        LearnBinarySearchAlgo game = new LearnBinarySearchAlgo(100);
        int result = game.guessNumber(100);

        System.out.println("Binary search found: " + result);
        System.out.println("Actual picked number: " + game.getPick());

        // ---- SEARCH IN ROTATED SORTED ARRAY ----
        int[] nums = {4,5,6,7,0,1,2};
        int target = 0;
        int index = search(nums, target);
        System.out.println("Array: [4,5,6,7,0,1,2]");
        System.out.println("Target " + target + " found at index: " + index);
    }
    // Square root of a number using binary search
    private static int sqrt(int n) {
        if (n < 2) return n; // Handle 0 and 1

        int left = 1, right = n / 2, ans = 0;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            long square = (long) mid * mid; // Use long to avoid overflow

            if (square == n) {
                return mid; // Perfect square
            } else if (square < n) {
                ans = mid; // Store the last valid answer
                left = mid + 1; // Search in the right half
            } else {
                right = mid - 1; // Search in the left half
            }
        }
        return ans; // Return the floor of the square root
    }
    // Complexity analysis:
    // Time Complexity: O(log n) - Binary search halves the search space with each iteration
    // Space Complexity: O(1) - Constant space used for variables
    // Explanation: The function uses binary search to find the largest integer whose square is less than or equal to n.
    // It maintains a search range defined by left and right pointers, and iteratively narrows down this range until it finds the correct integer.

    // Guess Number Game using binary search
    // The guess API is defined in the parent class. We are trying to guess a number between 1 and n.
    public int guessNumber(int n) {
        int left = 1, right = n;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            int res = guess(mid);

            if (res == 0) { 
                return mid; // Found the correct number
            } else if (res < 0) {
                right = mid - 1; // Guess is too high, search in the left half
            } else {
                left = mid + 1; // Guess is too low, search in the right half
            } 
        }
        return -1;
    }
    // Complexity analysis:
    // Time Complexity: O(log n) - Binary search halves the search space with each iteration
    // Space Complexity: O(1) - Constant space used for variables
    // Explanation: The function uses binary search to find the picked number. '
    // It maintains a search range defined by left and right pointers, and iteratively narrows down this range based on the feedback from the guess API until it finds the correct number.

    // Search in a rotated sorted array using binary search
    // The array is sorted in ascending order and then rotated at some pivot. We need to find the index of the target value.
    public static int search(int[] nums, int target) {
        int l = 0, r = nums.length - 1 ;
        while( l <=r ){
            int m = l + (r-l)/2;
            if(target == nums[m]){
                return m;
            }

            if(nums[l] <= nums[m]){ // This checks if the left half of the array (from index l to m) is sorted.
                if(target < nums[m] && target >= nums[l]){ 
                    // This checks if the target is within the sorted left half. 
                    // If it is, we can safely ignore the right half and continue searching in the left half by moving the right pointer to m - 1.
                    r = m - 1;
                }else{
                    l = m + 1; // If the target is not in the sorted left half, it must be in the right half, so we move the left pointer to m + 1 to continue searching there.
                }
            }else{ // right side sorted 
                if(target > nums[m] && target <= nums[r]){
                    // This checks if the target is within the sorted right half. 
                    // If it is, we can safely ignore the left half and continue searching in the right half by moving the left pointer to m + 1.
                    l = m + 1;
                }else{
                    r = m - 1; // If the target is not in the sorted right half, it must be in the left half, so we move the right pointer to m - 1 to continue searching there.
                }                
            }
        }

        return -1;
    }
    // Complexity analysis:
    // Time Complexity: O(log n) - Binary search halves the search space with each iteration
    // Space Complexity: O(1) - Constant space used for variables
    // Explanation: The function searches for a target value in a rotated sorted array. 
    // It uses binary search to efficiently find the target by determining which half of the array is sorted and adjusting the search range accordingly.

    public static int firstBadVersion(int n) {
        int l = 1 ,r = n;
        while(l < r){ // 
            int m = l + (r-l)/2;
            if(!isBadVersion(m)){ 
                l = m + 1;
            }else{
                r = m;
            }
        }
        return r ;
    }
    // Complexity analysis:
    // Time Complexity: O(log n) - Binary search halves the search space with each iteration
    // Space Complexity: O(1) - Constant space used for variables
    // Explanation: The function finds the first bad version among n versions.
    // It uses binary search to efficiently find the first bad version by checking the middle version
    // and adjusting the search range based on whether the middle version is bad or not.
    // The loop continues until the search range is narrowed down to a single version, which is the first bad version.
}

File: JavascriptPrograms/LearnBinarySearchAlgo.js
// ---- SQUARE ROOT OF A NUMBER ----

/**
 * @param {number} x
 * @return {number}
 */
var mySqrt = function (x) {
  if (x < 2) return x;

  let l = 2;
  let r = Math.floor(x / 2);

  while (l <= r) {
    let m = l + Math.floor((r - l) / 2);
    if (x === m * m) {
      return m;
    } else if (x < m * m) {
      r = m - 1;
    } else {
      l = m + 1;
    }
  }

  return r;
};
console.log(mySqrt(8)); // Output: 2
console.log(mySqrt(4)); // Output: 2

// Complexity Analysis
// Time complexity : O(logn). Binary search takes logarithmic time.
// Space complexity : O(1). Constant space is used.
// Explanation of the code
// We start with two pointers, l and r, initialized to 2 and x/2 respectively.
// We perform binary search by calculating the middle point m.
// If m*m equals x, we have found the square root and return m.
// If m*m is greater than x, we adjust the right pointer r to m-1.
// If m*m is less than x, we adjust the left pointer l to m+1.
// The loop continues until l exceeds r.
// Finally, we return r, which will be the integer part of the square root of x.
// Examples
// For x = 8:
// Initial l = 2, r = 4
// 1st iteration: m = 3, 3*3 = 9 > 8, so r = 2
// 2nd iteration: m = 2, 2*2 = 4 < 8, so l = 3
// Now l > r, we return r = 2.
// For x = 4:
// Initial l = 2, r = 2
// 1st iteration: m = 2, 2*2 = 4 == 4, we return m = 2.
// The code correctly calculates the integer square root of a non-negative integer x using binary search.

// ---- GUESS NUMBER HIGHER OR LOWER ----

/**
 * Forward declaration of guess API.
 * @param {number} num   your guess
 * @return 	     -1 if num is higher than the picked number
 *			      1 if num is lower than the picked number
 *               otherwise return 0
 * var guess = function(num) {}
 */

/**
 * @param {number} n
 * @return {number}
 */
var guessNumber = function (n) {
  let l = 1;
  let r = n;

  while (l <= r) {
    let m = l + Math.floor((r - l) / 2);
    let result = guess(m);
    if (result === 0) {
      return m;
    } else if (result < 0) {
      r = m - 1;
    } else if (result > 0) {
      l = m + 1;
    }
  }
  return -1;
};

console.log(guessNumber(10)); // Output: 6 (assuming the picked number is 6)
// Complexity Analysis
// Time complexity : O(logn). Binary search takes logarithmic time.
// Space complexity : O(1). Constant space is used.
// Explanation of the code
// We start with two pointers, l and r, initialized to 1 and n respectively.
// We perform binary search by calculating the middle point m.
// We call the guess API with m and store the result.
// If the result is 0, we have found the picked number and return m.
// If the result is -1, it means m is higher than the picked number, so we adjust the right pointer r to m-1.
// If the result is 1, it means m is lower than the picked number, so we adjust the left pointer l to m+1.
// The loop continues until l exceeds r.
// If we exit the loop without finding the picked number, we return -1.
// The code correctly implements a binary search to find the picked number using the guess API.

// ---- SEARCH IN ROTATED SORTED ARRAY ----

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function (nums, target) {
  let l = 0;
  let r = nums.length - 1;
  while (l <= r) {
    let m = l + Math.floor((r - l) / 2);
    if (target == nums[m]) {
      return m;
    }

    if (nums[l] <= nums[m]) {
      // left side sorted
      if (target < nums[m] && target >= nums[l]) {
        r = m - 1;
      } else {
        l = m + 1;
      }
    } else {
      if (target > nums[m] && target <= nums[r]) {
        l = m + 1;
      } else {
        r = m - 1;
      }
    }
  }
  return -1;
};
console.log(search([4, 5, 6, 7, 0, 1, 2], 0)); // Output: 4
// Complexity Analysis
// Time complexity : O(logn). Binary search takes logarithmic time.
// Space complexity : O(1). Constant space is used.
// Explanation of the code
// We start with two pointers, l and r, initialized to 0 and the last index of the array respectively.
// We perform binary search by calculating the middle point m.
// If the target is found at index m, we return m.
// We check if the left side of the array (from l to m) is sorted by comparing nums[l] and nums[m].
// If the left side is sorted, we check if the target is within the range of the left side. If it is, we adjust the right pointer r to m-1. Otherwise, we adjust the left pointer l to m+1.
// If the left side is not sorted, it means the right side (from m to r) is sorted. We check if the target is within the range of the right side. If it is, we adjust the left pointer l to m+1. Otherwise, we adjust the right pointer r to m-1.
// The loop continues until l exceeds r.
// If we exit the loop without finding the target, we return -1.
// The code correctly implements a binary search to find the target in a rotated sorted array.

// ---- FIRST BAD VERSION ----
/**
 * Definition for isBadVersion()
 *
 * @param {integer} version number
 * @return {boolean} whether the version is bad
 * isBadVersion = function(version) {
 *     ...
 * };
 */
var isBadVersion = function (version) {
  // This is a placeholder for the actual implementation of the isBadVersion API.
  // In a real scenario, this would be provided by the system or defined elsewhere in the codebase.
  return version >= 4; // Example: versions 4 and above are bad
};
/**
 * @param {function} isBadVersion()
 * @return {function}
 */
var solution = function (isBadVersion) {
  /**
   * @param {integer} n Total versions
   * @return {integer} The first bad version
   */
  return function (n) {
    let l = 1;
    let r = n;

    while (l < r) {
      let m = l + Math.floor((r - l) / 2);
      if (!isBadVersion(m)) {
        l = m + 1;
      } else {
        r = m;
      }
    }

    return r;
  };
};
console.log(solution(isBadVersion)(5)); // Output: 4
// Complexity Analysis
// Time complexity : O(logn). Binary search takes logarithmic time.
// Space complexity : O(1). Constant space is used.
// Explanation of the code
// We start with two pointers, l and r, initialized to 1 and n respectively.
// We perform binary search by calculating the middle point m.
// We call the isBadVersion API with m to check if it is a bad version.
// If m is not a bad version, it means the first bad version must be after m, so we adjust the left pointer l to m+1.
// If m is a bad version, it means the first bad version is at m or before m, so we adjust the right pointer r to m.
// The loop continues until l is less than r.
// Finally, we return r, which will be the first bad version.
// The code correctly implements a binary search to find the first bad version using the isBadVersion API.
